### PR TITLE
Merge emacs 2018-04-01.

### PR DESCRIPTION
**Pull request by shaleh**

### CODE REVIEW

**Comment by brotzeit** [comment]
FAILED  eval-tests--catch-base
   FAILED  eval-tests--funcall-error
   FAILED  eval-tests--function-base

Hmm. Any idea ?

**Comment by shaleh** [comment]
Not yet, I have not had a chance to look at the errors.

**Comment by harryfei** [comment]
Actually, the failed tests run successfully in my local machine.

**Comment by shaleh** [comment]
Ok @harryfei I just rebased this using the web interface. Let's see how it goes...

**Comment by daichimukai** [comment]
I got errors same as in @brotzeit's comment on my machine. When I run a test manually I see

```
$ ../src/remacs -batch --no-site-lisp --no-site-file -L :. -l ert -l rust_src/src/eval-tests --eval "(ert-run-tests-batch-and-exit (quote (not (or (tag :expensive-test) (tag :unstable)))))"
```

has failed, but

```
$ ../src/remacs -batch --no-site-lisp --no-site-file -L :. -l ert -l rust_src/src/eval-tests.el --eval "(ert-run-tests-batch-and-exit (quote (not (or (tag :expensive-test) (tag :unstable)))))"
```
has finished successfully. Only the difference is whether I added `.el` or not. `signal` or related codes seems to be not good with byte compiling, but I don't know actually what's happening on my machine.'

The log file of failed test is [here](https://gist.github.com/daichimukai/b60c1fefeacd0af64be6e324ac2293df).

**Comment by harryfei** [comment]
I used the command `make rust_src/src/eval-tests`. It has finished successfully.

**Comment by shaleh** [comment]
Need to re-run due to the project move.

**Comment by daichimukai** [comment]
@harryfei Yeah, the same for me. Could you check `make TEST_LOAD_EL=no rust_src/src/eval-tests`? This fails for me.

File: rust_src/src/data.rs [hunk 1]
```diff
@@ -844,7 +844,7 @@ pub fn string_to_number_lisp(mut string: LispStringRef, base: Option<EmacsInt>)
         }
     }
 
-    match unsafe { string_to_number(p, b as i32, true) } {
+    match unsafe { string_to_number(p, b as i32, 0) } {
```
Comment (daichimukai) [inline rust_src/src/data.rs]: `S2N_IGNORE_TRAILING` in lisp.h, which is 1 in these code, seems to be an alternative for `true`.

Comment (shaleh) [reply rust_src/src/data.rs]: I just looked at the emacs source and it is 0. do you see something different?

Comment (daichimukai) [reply rust_src/src/data.rs]: I'm seeing this. I was wrong, I should suggest `I2N_IGNORE_TRAILING | S2N_OVERFLOW_TO_NUMBER`.

https://github.com/remacs/remacs/blob/7bedc8812bd7ca1d9cf36636322068b28b690a85/src/data.c#L2737-L2738

**Comment by harryfei** [comment]
@daichimukai I can get the same error now.
@shaleh If I use `make check`, I can reproduce the bug in CI.

**Comment by brotzeit** [comment]
It seems we have to handle functions, that behave differently when called noninteractively, seperately. I would suggest we add an additional directory and call those tests manually after regular make check.

File: rust_src/src/marker.rs [hunk 2]
```diff
@@ -603,6 +611,26 @@ impl LispBufferRef {
...
+// The previous code used INITIAL=50 and INCREMENT=0 and this lead to
+// really poor performance when there are many markers.
+// I haven't tried to tweak INITIAL, but experiments on my trusty Thinkpad
+// T61 using various artificial test cases seem to suggest that INCREMENT=50
+// might be "the best compromise": it significantly improved the
+// worst case and it was rarely slower and never by much.
+// The asymptotic behavior is still poor, tho, so in largish buffers with many
+// overlays (e.g. 300KB and 30K overlays), it can still be a bottleneck.
+static BYTECHAR_DISTANCE_INITIAL: isize = 50;
+static BYTECHAR_DISTANCE_INCREMENT: isize = 50;
```
Comment (brotzeit) [inline rust_src/src/marker.rs]: Where do we need this ?

Comment (agraven) [reply rust_src/src/marker.rs]: I guess it's intended to be used somewhere in `LispBufferRef::charpos_to_bytepos`, but isn't.

**Comment by shaleh** [comment]
I suspect the best thing to do is break this PR down into a week at a time. As I recall the bug prevent CI to pass happens towards the end of the first week.

**Comment by shanavas786** [comment]
#1388 handles a smaller piece

**Comment by shanavas786** [comment]
Some failing tests pass when run interactively.

**Comment by shaleh** [comment]
Yep. The bug starts around day 4 of the merge I think.

**Comment by brotzeit** [comment]
There's a lot of strange stuff happening when running tests non-interactively. I'm adding tests to the libvterm stuff and the prompt gets split into multiple lines and I have no idea why.

**Comment by brotzeit** [comment]
I just realized the window size is extremely small...

**Comment by shaleh** [comment]
Aborting this one.

**Pull request closed without merging**